File: Sums4_part2/Sum_4_part2.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

public class Main {
    public static void main(String[] args) {
        process(new int[] {1, 2}, new int[] {-2, -1},
                new int[] {-1, 2}, new int[] {0, 2});
        process(new int[] {0}, new int[] {0},
                new int[] {0}, new int[] {0});

    }

    private static void process(int[] list1, int[] list2, int[] list3, int[] list4) {
        ArrayList<ArrayList<Integer>> group = new ArrayList<>();
        ArrayList<ArrayList<Integer>> index_tuple = new ArrayList<>();


        for (int i=0; i<list1.length; i++) {
            for (int j=0; j<list2.length; j++) {
                for (int k=0; k<list3.length; k++) {
                    for (int l=0; l<list4.length; l++) {
                        if (list1[i] + list2[j] + list3[k] + list4[l] == 0 ) {
                            ArrayList<Integer> result = new ArrayList<>();
                            ArrayList<Integer> result_tuple = new ArrayList<>();
                            result.add(list1[i]);
                            result.add(list2[j]);
                            result.add(list3[k]);
                            result.add(list4[l]);
                            result_tuple.add(i);
                            result_tuple.add(j);
                            result_tuple.add(k);
                            result_tuple.add(l);
                            Collections.sort(result);
                           
                            if (!group.contains(result)) {
                                group.add(result);
                            }
                            if (!index_tuple.contains(result_tuple)) {
                                index_tuple.add(result_tuple);
                            }
                        }

                    }
                }
            }
        }

File: Sums4_part2/Sum_4_part2.py
def sum_4_part2(list1, list2, list3, list4):
    group = []
    index_tuple = []

    for i in range(len(list1)):
        for j in range(len(list2)):
            for k in range(len(list3)):
                for l in range(len(list4)):
                    if list1[i] + list2[j] + list3[k] + list4[l] == 0:
                        result = sorted([list1[i], list2[j], list3[k], list4[l]])
                        result_tuple = (i, j, k, l)
                        if result not in group:
                            group.append(result)
                        if result_tuple not in index_tuple:
                            index_tuple.append(result_tuple)

    print(f"Number lists whose sum is zero = {group}")
    print(f"Indexes tuples of numbers whose sum is zero =  {index_tuple}")
    length = len(index_tuple)
    print(f"Length of the tuple list = {length}")

sum_4_part2([1, 2],[-2, -1], [-1, 2], [0, 2])
sum_4_part2([0],[0],[0],[0])
